### PR TITLE
[SPARK-27605][UI] Add new column "Partition ID" to the tasks table in stages page

### PR DESCRIPTION
**Pull request by pgandhi999**
If you have more than one stage attempt in a Spark job, the task index will not equal partition id. Thus, adding a column for partition id would be a useful feature to have.

## What changes were proposed in this pull request?

To get the data for partition id in UI, it has to be logged into the event files while the job is running. So, the code changes mainly focus on that. Also, backward compatibility with old history files has been ensured as well here.

## How was this patch tested?

Screen Shot 2019-04-30 at 9 13 58 AM

### CODE REVIEW

**Comment by pgandhi999** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/status/storeTypes.scala [hunk 1]
```diff
@@ -141,6 +141,7 @@ private[spark] object TaskIndexNames {
   final val STATUS = "sta"
   final val TASK_INDEX = "idx"
   final val COMPLETION_TIME = "ct"
+  final val PARTITION_ID = "pid"
```
Comment (redsanket) [inline core/src/main/scala/org/apache/spark/status/storeTypes.scala]: nit: "pid" seems to be more in terms of process id, may be possible to change that?

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 2]
```diff
@@ -304,7 +304,8 @@ private[spark] object JsonProtocol {
     ("Finish Time" -> taskInfo.finishTime) ~
     ("Failed" -> taskInfo.failed) ~
     ("Killed" -> taskInfo.killed) ~
-    ("Accumulables" -> accumulablesToJson(taskInfo.accumulables))
+    ("Accumulables" -> accumulablesToJson(taskInfo.accumulables)) ~
+    ("Partition ID" -> taskInfo.partitionId)
```
Comment (redsanket) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: nit: Not sure if it is ideal to maintain the order of the UI display like follow Partition ID follows Task Indec?

File: core/src/test/resources/HistoryServerExpectations/blacklisting_for_stage_expectation.json [hunk 3]
```diff
@@ -134,7 +135,8 @@
         "stderr" : "http://172.30.65.138:64278/logPage/?appId=app-20180109111548-0000&executorId=1&logType=stderr"
       },
       "schedulerDelay" : 10,
-      "gettingResultTime" : 0
+      "gettingResultTime" : 0,
+      "partitionId" : -1
```
Comment (redsanket) [inline core/src/test/resources/HistoryServerExpectations/blacklisting_for_stage_expectation.json]: I see partitionId as -1 in the test cases everywhere, does this need some test case where it is >= 0?

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 4]
```diff
@@ -804,9 +805,11 @@ private[spark] object JsonProtocol {
       case Some(values) => values.map(accumulableInfoFromJson)
       case None => Seq.empty[AccumulableInfo]
     }
+    val partitionId = jsonOption(json \ "Partition ID").map(_.extract[Int]).getOrElse(-1)
 
     val taskInfo =
       new TaskInfo(taskId, index, attempt, launchTime, executorId, host, taskLocality, speculative)
+    taskInfo.setPartitionId(partitionId)
```
Comment (redsanket) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: do we need a setting other vars like gettingResultTime and finishTime seem to be accessed 
taskInfo.gettingResultTime = gettingResultTime
taskInfo.finishTime = finishTime

May be make it consistent if we need we want setters and make it private for others as well?

File: core/src/main/resources/org/apache/spark/ui/static/stagepage.js [hunk 5]
```diff
@@ -933,18 +942,18 @@ $(document).ready(function () {
...
                 // summary table
-                taskTableSelector.column(19).visible(dataToShow.showInputData);
-                taskTableSelector.column(20).visible(dataToShow.showOutputData);
-                taskTableSelector.column(21).visible(dataToShow.showShuffleWriteData);
+                taskTableSelector.column(20).visible(dataToShow.showInputData);
+                taskTableSelector.column(21).visible(dataToShow.showOutputData);
                 taskTableSelector.column(22).visible(dataToShow.showShuffleWriteData);
-                taskTableSelector.column(23).visible(dataToShow.showShuffleReadData);
-                taskTableSelector.column(24).visible(dataToShow.showBytesSpilledData);
+                taskTableSelector.column(23).visible(dataToShow.showShuffleWriteData);
```
Comment (dhruve) [inline core/src/main/resources/org/apache/spark/ui/static/stagepage.js]: Why do we have this repeated twice?

Comment (pgandhi999) [reply core/src/main/resources/org/apache/spark/ui/static/stagepage.js]: It was like this before. I have simply changed the column numbers. The flag showShuffleWriteData applies to both the columns "Write Time" as well as "Shuffle Write Size/Records".

**Comment by pgandhi999** [comment]
test this please.

**Comment by pgandhi999** [comment]
ok to test

**Pull request closed without merging**